core/src/task.rs

```diff
@@ -533,48 +537,151 @@ macro_rules! vdaf_dispatch {
-/// collector-to-aggregator authentication.
-#[derive(Clone)]
-pub struct AuthenticationToken(Vec<u8>);
+/// Different modes of authentication supported by Janus for either sending requests (e.g., leader
+/// to helper) or receiving them (e.g., collector to leader).
+#[derive(Clone, Derivative, Serialize, Deserialize)]
+#[derivative(Debug)]
+#[serde(tag = "type", content = "token")]
+#[non_exhaustive]
+pub enum AuthenticationToken {
+    /// A bearer token. The value is an opaque byte string. Its Base64 encoding is inserted into
+    /// HTTP requests as specified in [RFC 6750 section 2.1][1]. The token is not necessarily an
+    /// OAuth token.
+    ///
+    /// [1]: https://datatracker.ietf.org/doc/html/rfc6750#section-2.1
+    Bearer(
+        #[derivative(Debug = "ignore")]
+        #[serde(
+            serialize_with = "as_url_safe_unpadded_base64",
+            deserialize_with = "from_url_safe_unpadded_base64"
+        )]
+        Vec<u8>,
+    ),
+
+    /// Token presented as the value of the "DAP-Auth-Token" HTTP header. Conforms to
+    /// [draft-ietf-dap-ppm-01 section 3.2][1].
+    ///
+    /// [1]: https://datatracker.ietf.org/doc/html/draft-ietf-ppm-dap-01#name-https-sender-authentication
+    DapAuth(DapAuthToken),
+}
 
 impl AuthenticationToken {
-    /// Constructs a bearer token string suitable for use as the value in an HTTP `Authorization`
-    /// header.
-    pub fn bearer_token(&self) -> String {
-        format!("Bearer {}", STANDARD.encode(self.as_ref()))
+    /// Returns an HTTP header and value that should be used to authenticate an HTTP request with
+    /// this credential.
+    pub fn request_authentication(&self) -> (&'static str, Vec<u8>) {
+        match self {
+            Self::Bearer(token) => (
+                AUTHORIZATION.as_str(),
+                // When encoding into a request, we use Base64 standard encoding
+                format!("Bearer {}", STANDARD.encode(token.as_slice())).into_bytes(),
+            ),
+            // A DAP-Auth-Token is already HTTP header-safe, so no encoding is needed. Cloning is
+            // unfortunate but necessary since other arms must allocate.
+            Self::DapAuth(token) => (DAP_AUTH_HEADER, token.as_ref().to_vec()),
+        }
     }
 }
+
+impl PartialEq for AuthenticationToken {
+    fn eq(&self, other: &Self) -> bool {
+        let (own, other) = match (self, other) {
+            (Self::Bearer(own), Self::Bearer(other)) => (own.as_slice(), other.as_slice()),
+            (Self::DapAuth(own), Self::DapAuth(other)) => (own.as_ref(), other.as_ref()),
+            _ => {
+                return false;
+            }
+        };
+        // We attempt constant-time comparisons of the token data. Note that this function still
+        // leaks whether the lengths of the tokens are equal -- this is acceptable because we expect
+        // the content of the tokens to provide enough randomness that needs to be guessed even if
+        // the length is known.
```

Comment:
```suggestion
        // We attempt constant-time comparisons of the token data to mitigate timing attacks. Note
        // that this function still leaks whether the lengths of the tokens are equal -- this is acceptable
        // because we expect the content of the tokens to provide enough randomness that needs to 
        // be guessed even if the length is known.
```

Optional: Adding "to mitigate timing attacks", prefer to be explicit about what is being accomplished.